library/core/src/ptr/mod.rs

```diff
@@ -230,7 +255,146 @@ pub const fn null<T>() -> *const T {
 #[rustc_const_stable(feature = "const_ptr_null", since = "1.24.0")]
 #[rustc_diagnostic_item = "ptr_null_mut"]
 pub const fn null_mut<T>() -> *mut T {
-    0 as *mut T
+    invalid_mut::<T>(0)
+}
+
+/// Forge a pointer to a Zero-Sized Type (ZST) from nothing.
+///
+/// Zero-sized types do not actually exist in memory, and therefore you cannot actually
+/// "read" or "write" a ZST (any method that claims to do so is just playing pretend,
+/// although you do still need to respect alignment for something like `&[u32; 0]`).
+/// As a result, you are free to claim a ZSTs exists anywhere you want (except null).
+///
+/// This API exists to make the soundness of this pattern explicit, even under
+/// "strict provenance". It is equivalent to the deprecated `addr as *mut T` cast.
+///
+/// **BUT YOU AREN'T ACTUALLY ALLOWED TO BLINDLY FORGE ZST _INSTANCES_.**
+///
+/// It's sound for an API to use an instance of a ZST to enforce some important
+/// safety property. So for instance, you can make an API like this:
+///
+/// ```ignore
+/// pub struct Step1Token(_private_to_construct: ());
+///
+/// pub fn step1() -> Step1Token { ... }
+/// pub fn step2(proof: Step1Token) { ... }
+/// ```
+///
+/// And it's sound in the body of `step2` to assume that `step1` has been run
+/// beforehand, because the only way to get an instance of Step1Token is to call
+/// `step1` (assuming `step1` is indeed the only API that creates one).
+///
+/// A well-behaved abstraction should conceptually only be "reading" ZSTs that it
+/// has previously "written". You don't *actually* need to do the write, and could
+/// feed it into [`mem::forget`][] instead, but for ZSTs `write` is a perfectly
+/// good way to say `forget` and better expresses the semantics of your code.
+///
+/// Anything that stores *many* ZSTs should at the minimum maintain a counter of how
+/// many it has written so that it can know how many it can/must read later.
+/// `Vec<()>` is basically just a counter that goes up on `push` and down on `pop`.
+///
+/// Note: if you need to "allocate" memory for a buffer of ZSTs,
+/// [`core::ptr::NonNull::dangling`][] is more useful, because it
+/// handles alignment for you.
+///
+/// # Example
+///
+/// ```
+/// use core::{ptr, mem};
+///
+/// // I store my ZSTs at the *coolest* address
+/// let my_good_ptr = ptr::zst_exists::<()>(0xc001_add7);
+///
+/// // "store" and then "load" a ZST at this cool address.
+/// my_good_ptr.write(());
+/// let output = my_good_ptr.read();
+/// ```
+#[inline(always)]
+#[must_use]
+#[rustc_promotable]
```

Comment:
I believe this is an historical accident and shouldn't be applied to more things.

Reply:
just the promotable?

Reply:
Yes

Reply:
Yes please no new `rustc_promotable` methods. :)